src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const clone = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        stateHistory.push({ ...addProperties(clone, action.extraData) });
+        break;
+
+      case 'removeProperties':
+        stateHistory.push({ ...removeProperties(clone, action.keysToRemove) });
+        break;
+
+      case 'clear':
+        stateHistory.push({ ...clearProperties(clone) });
+        break;
+    }
+  }
```

Comment:
[DONT REPEAT YOURSELF] - Pushing to 'stateHistory' is an action performed in all cases of the switch. Consider doing it just once after the switch block to avoid repetition.